diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/NestedFunctionBlock.kt

```diff
@@ -0,0 +1,106 @@
+import org.jetbrains.kotlin.psi.KtClass
+
+/**
+ * Rule 5.1.2 Nested blokcs
+ */
+class NestedFunctionBlock(private val configRules: List<RulesConfig>) : Rule("nested-block") {
+
+    companion object {
+        private const val MAX_NESTED_BLOCK_COUNT = 4L
+    }
+
+    val configuration: NestedBlockConfiguration by lazy {
+        NestedBlockConfiguration(configRules.getRuleConfig(NESTED_BLOCK)?.configuration ?: mapOf())
+    }
+
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        when (node.elementType) {
+            FUN, FUNCTION_LITERAL -> handleFunForNestedBlocks(node, configuration.maxNestedBlockQuantity)
+            CLASS, OBJECT_DECLARATION -> handleLocalCLassForNestedBlocks(node, configuration.maxNestedBlockQuantity)
+        }
+    }
+
+    private fun handleLocalCLassForNestedBlocks(node: ASTNode, maxNestedBlockCount: Long) {
+        if (node.elementType == CLASS && !(node.psi as KtClass).isLocal) return
+        node.getChildren(TokenSet.create(CLASS_BODY)).forEach {
+            dfsBlock(it, node, 1, maxNestedBlockCount)
+        }
+    }
+
+    private fun handleFunForNestedBlocks(node: ASTNode, maxNestedBlockCount: Long) {
+        val classParentNode = node.parent({ it.elementType == CLASS })
+        if (classParentNode != null && (classParentNode.psi as KtClass).isLocal) return
+        node.getChildren(TokenSet.create(BLOCK)).forEach {
+            dfsBlock(it, node, 1, maxNestedBlockCount)
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun dfsBlock(node: ASTNode, initialNode: ASTNode, blockCount: Int, maxNestedBlockCount: Long) {
+        if (blockCount > maxNestedBlockCount) {
+            NESTED_BLOCK.warn(configRules, emitWarn, isFixMode, initialNode.findChildByType(IDENTIFIER)?.text
+                    ?: initialNode.text,
+                    initialNode.startOffset, initialNode)
+            return
+        } else {
+            findBlocks(node).forEach {
+                dfsBlock(it, initialNode, blockCount + 1, maxNestedBlockCount)
+            }
+        }
+    }
+
+    private fun findBlocks(node: ASTNode): List<ASTNode> {
+        val result = mutableListOf<ASTNode>()
+        node.getChildren(null).forEach { childNode ->
+            when (childNode.elementType) {
+                IF -> Pair(childNode.findChildByType(THEN)?.findChildByType(BLOCK),
+                        (childNode.findChildByType(ELSE)?.findChildByType(BLOCK)
+                                ?: if(childNode.findChildByType(ELSE)?.findChildByType(IF) != null)
+                                    childNode.findChildByType(ELSE) else null))
```

Comment:
```suggestion
                                ?: childNode.findChildByType(ELSE)?.takeIf { it.findChildByType(IF) != null })
```
just to reduce amount of conditional operators